src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,183 @@
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> addedNode = new Node<>(tail, value, null);
+        if (size > 0) {
+            tail.next = addedNode;
+        } else {
+            head = addedNode;
+        }
+        tail = addedNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkPositionIndex(index);
+        if (index == 0) {
+            Node<T> addedNode = new Node<>(null, value, head);
+            head.previous = addedNode;
+            head = addedNode;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> prevNode = currentNode.previous;
+            Node<T> addedNode = new Node<>(prevNode, value, currentNode);
+            currentNode.previous = addedNode;
+            prevNode.next = addedNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkPositionIndex(index);
+        return node(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> currentNode = node(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkPositionIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> node = head; node != null; node = node.next) {
+                if (node.value == null) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> currentNode = head; currentNode != null; currentNode = currentNode.next) {
+                if (object.equals(currentNode.value)) {
+                    unlink(currentNode);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + ", Size: " + size;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private Node<T> node(int index) {
```

Comment:
uninformative method name, and you already have `getNode(int index)` method, what the difference?